### PR TITLE
Add tool support for jlama

### LINKED ISSUE #1660: [FEATURE] Jlama: Add Tool Support
Update Jlama artifact and integrate tool calling into langchain4j integration

### PR DESCRIPTION
**Pull request by tjake**
## Issue

Closes #1660 

## Change
This PR adds tool calling support for Jlama added in 0.3.1 release.

I have tested these changes with Mistral and Llama-3.1 models.

## General checklist

- [X] There are no breaking changes
- [X] I have added unit and integration tests for my change
- [X] I have manually run all the unit and integration tests in the module I have added/changed, and they are all green
- [X] I have manually run all the unit and integration tests in the [core](https://github.com/langchain4j/langchain4j/tree/main/langchain4j-core) and [main](https://github.com/langchain4j/langchain4j/tree/main/langchain4j) modules, and they are all green

- [X] I have added/updated the [documentation](https://github.com/langchain4j/langchain4j/tree/main/docs/docs)
- [X] I have added an example in the [examples repo](https://github.com/langchain4j/langchain4j-examples) (only for "big" features)
- [ ] I have added/updated [Spring Boot starter(s)](https://github.com/langchain4j/langchain4j-spring) (if applicable)

## Checklist for adding new model integration

- [ ] I have added my new module in the [BOM](https://github.com/langchain4j/langchain4j/blob/main/langchain4j-bom/pom.xml)

## Checklist for adding new embedding store integration

- [ ] I have added a `{NameOfIntegration}EmbeddingStoreIT` that extends from either `EmbeddingStoreIT` or `EmbeddingStoreWithFilteringIT`
- [ ] I have added my new module in the [BOM](https://github.com/langchain4j/langchain4j/blob/main/langchain4j-bom/pom.xml)

## Checklist for changing existing embedding store integration

- [ ] I have manually verified that the `{NameOfIntegration}EmbeddingStore` works correctly with the data persisted using the latest released version of LangChain4j

### CODE REVIEW

**Comment by tjake** [comment]
Examples PR: https://github.com/langchain4j/langchain4j-examples/pull/106

**Comment by tjake** [comment]
This release also includes a fix for https://github.com/langchain4j/langchain4j/issues/1506

File: langchain4j-jlama/src/main/java/dev/langchain4j/model/jlama/JlamaChatModel.java [hunk 1]
```diff
@@ -49,7 +52,7 @@ public JlamaChatModel(Path modelCachePath,
             loader = loader.workingDirectory(workingDirectory);
 
         this.model = loader.load();
-        this.temperature = temperature == null ? 0.7f : temperature;
+        this.temperature = temperature == null ? 0.3f : temperature;
```
Comment (langchain4j) [inline langchain4j-jlama/src/main/java/dev/langchain4j/model/jlama/JlamaChatModel.java:52]: Is this change intentional?

File: langchain4j-jlama/src/main/java/dev/langchain4j/model/jlama/JlamaChatModel.java [hunk 2]
```diff
@@ -62,22 +65,68 @@ public static JlamaChatModelBuilder builder() {
...
-                case USER -> promptBuilder.addUserMessage(message.text());
-                case AI -> promptBuilder.addAssistantMessage(message.text());
+                case SYSTEM -> promptBuilder.addSystemMessage(((SystemMessage)message).text());
+                case USER -> {
+                    UserMessage userMessage = (UserMessage)message;
+                    for (Content content : userMessage.contents()) {
+                        if (content.type() != ContentType.TEXT)
+                            throw new UnsupportedOperationException("Unsupported content type: " + content.type());
+
+                        promptBuilder.addUserMessage(((TextContent)content).text());
```
Comment (langchain4j) [inline langchain4j-jlama/src/main/java/dev/langchain4j/model/jlama/JlamaChatModel.java]: I am not sure this is an intended behavior. I guess all text contents should be concatenated together into a single user message?

**Review by ghost — COMMENTED** [review]
@tjake thank you!

**Comment by langchain4j** [comment]
@tjake those 2 comments are not critical, if you won't be able to address them till wednesday, I will merge it anyway to include it in this release

File: langchain4j-jlama/src/main/java/dev/langchain4j/model/jlama/JlamaChatModel.java (hunk 1 above)
Comment (tjake) [reply langchain4j-jlama/src/main/java/dev/langchain4j/model/jlama/JlamaChatModel.java:52]: Yeah, since jlama 3.1 does badly with > 0.3 according to them and it's ok as a default IMO

File: langchain4j-jlama/src/main/java/dev/langchain4j/model/jlama/JlamaChatModel.java (hunk 2 above)
Comment (tjake) [reply langchain4j-jlama/src/main/java/dev/langchain4j/model/jlama/JlamaChatModel.java]: The prompt templates handle this by doing just that but it's upto the model?

Comment (langchain4j) [reply langchain4j-jlama/src/main/java/dev/langchain4j/model/jlama/JlamaChatModel.java]: This code:
```Java
UserMessage userMessage = UserMessage.from(List.of(TextContent.from("hello"), TextContent.from("world")));
model.generate(List.of(userMessage));
```
will create this prompt:
```
<|start_header_id|>system<|end_header_id|>

Environment: ipython
Cutting Knowledge Date: December 2023
Today Date: 26 Jul 2024

<|eot_id|><|start_header_id|>user<|end_header_id|>

Given the following functions, please respond with a JSON for a function call with its proper arguments that best answers the given prompt.

Respond in the format {"name": function name, "parameters": dictionary of argument name and its value}.Do not use variables.

hello<|eot_id|><|start_header_id|>user<|end_header_id|>

world<|eot_id|><|start_header_id|>assistant<|end_header_id|>

```

It seems there are 2 problems:
- 2 user messages instead of 1. I am not sure how exactly other LLM providers handle multiple contents in a single user message, but I guess some form of concatenation hapens. Usually contents are of different type though (e.g. text + image). So this is not really critical, can be reconsidered when/if other content types will be suported.
- Instructions regarding tools are present, but I did not specify any tools, Is this intended?

Comment (tjake) [reply langchain4j-jlama/src/main/java/dev/langchain4j/model/jlama/JlamaChatModel.java]: Hmm good catch I pushed a fix for second point.

Regarding the first point the order matters.  Esp for message history being passed in. Meaning I can't concat all user messages together arbitrarily.   
I think you're suggesting I collapse any messages of the same role that are sequential but I think the llm would handle it as is.

Comment (langchain4j) [reply langchain4j-jlama/src/main/java/dev/langchain4j/model/jlama/JlamaChatModel.java]: > Meaning I can't concat all user messages together arbitrarily.

I do not suggest concatenating all user messages together. I suggest concatenating all **_parts_** (contents) of a **_single_** user message together. In some LLM providers (e.g. OpenAI, Google Gemini) user message can consists of multiple parts (usually to support multimodal inputs), so `UserMessage` in LC4j consists of a `List<Content> contents`.

Comment (tjake) [reply langchain4j-jlama/src/main/java/dev/langchain4j/model/jlama/JlamaChatModel.java]: Sorry I get it now. fixed.

**Review by ghost — APPROVED** [review]
@tjake thanks a lot, good job!
